pkg/alerter/alerter.go

```diff
@@ -0,0 +1,245 @@
+		}
+		if n := len(errs); n > 0 {
+			resultCh <- fmt.Errorf("failed to update %d dashboards: %v", n, strings.Join(errs, ", "))
+		}
+		resultCh <- nil
+		close(resultCh)
+	}()
+
+	// Queue up all the dashboards
+	for _, d := range cfg.Dashboards {
+		if dashboard != "" && dashboard != d.Name {
+			logrus.WithField("dashboard", d.Name).Info("Skipping")
+			continue
+		}
+		dashboards <- d
+	}
+	close(dashboards)
+	wg.Wait() //wait for them to stop reporting
+	close(errCh)
+	return <-resultCh
+}
+
+// Paths to relevant GCS files
+var (
+	normalizer = regexp.MustCompile(`[^a-z0-9]+`)
+)
+
+// A normalized path for a summary file is "prefix-dashboardnameinlowercase"
+func summaryPath(name string) string {
+	return "summary-" + normalizer.ReplaceAllString(strings.ToLower(name), "")
+}
+
+func alerterPath(name string) string {
+	return "alerter-" + normalizer.ReplaceAllString(strings.ToLower(name), "")
+}
+
+// TODO(chases2): Move this and WriteSummary into a separate file in summarizer package
+func ReadSummary(ctx context.Context, client *storage.Client, path gcs.Path) (*summarypb.DashboardSummary, error) {
+	r, err := client.Bucket(path.Bucket()).Object(path.Object()).NewReader(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open config: %v", err)
+	}
+	buf, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read config: %v", err)
+	}
+	var summary summarypb.DashboardSummary
+	if err = proto.Unmarshal(buf, &summary); err != nil {
+		return nil, fmt.Errorf("failed to parse: %v", err)
+	}
+	return &summary, nil
+}
+
+func callOutDifferences(ctx context.Context, new, old *summarypb.DashboardSummary, config *configpb.Dashboard, subscribers []Klaxon) {
+	// Deserialize dashboard summaries into maps for easy access
+	oldSummaries := map[string]*summarypb.DashboardTabSummary{}
+	newSummaries := map[string]*summarypb.DashboardTabSummary{}
+
+	for _, tabSummary := range old.TabSummaries {
+		oldSummaries[tabSummary.DashboardTabName] = tabSummary
+	}
+
+	for _, tabSummary := range new.TabSummaries {
+		newSummaries[tabSummary.DashboardTabName] = tabSummary
+	}
+
+	for _, dashboardTab := range config.DashboardTab {
+		newSummary := newSummaries[dashboardTab.Name]
+		oldSummary := oldSummaries[dashboardTab.Name]
+
+		switch newSummary.OverallStatus {
+		case summarypb.DashboardTabSummary_PASS:
+			for _, sub := range subscribers {
+				err := sub.OnEachSuccess(ctx, newSummary, oldSummary, dashboardTab)
+				if err != nil {
+					logrus.Errorf("OnEachSuccess Dashboard %s Error: %v", dashboardTab.Name, err)
+				}
+			}
+		case summarypb.DashboardTabSummary_FAIL:
+			for _, sub := range subscribers {
+				err := sub.OnEachFailure(ctx, newSummary, oldSummary, dashboardTab)
+				if err != nil {
+					logrus.Errorf("OnEachFailure Dashboard %s Error: %v", dashboardTab.Name, err)
+				}
+			}
+		case summarypb.DashboardTabSummary_STALE:
+			if oldSummary.OverallStatus != summarypb.DashboardTabSummary_STALE {
```

Comment:
I do not see the value of this interface. It replaces an if statement with an OnBecomingStale method. The interface adds more complexity than it solves.

Reply:
Removed; now using a function injected to decouple. I think this does simplify things quite a lot, and should be easier to expand